### PR TITLE
Matrix improvements

### LINKED ISSUE #1431: Recover sender address in SignedMessage's from_dict methods
## Problem Definition

`SignedMessage` and its subclasses have a `decode()` method that is used to decode a packed binary representation of the message (by using the subclasses `unpack` methods). It sets the `sender` field by recovering the address from the message signature.

This doesn't happen when using `from_dict()` as it's used directly and not through a `decode()` like wrapper mechanism.

## Tasks

- [ ] Implement a mechanism for recovering the sender from the message signature when using `from_dict()`
- [ ] Update Matrix transport to use the new mechanism

### PR DESCRIPTION
**Pull request by andrevmatos**
- Better validation of signatures, uses `eth_sign` format/hasher
- Check for room parameters before joining, to avoid joining invalid/multiple rooms
- Makes `sender` a cached property calculated from signature. Fix #1431

### CODE REVIEW

File: raiden/messages.py [hunk 1]
```diff
@@ -1,15 +1,16 @@
-# -*- coding: utf-8 -*-
 from eth_utils import (
     big_endian_to_int,
-    encode_hex,
     to_normalized_address,
     to_canonical_address,
+    encode_hex,
+    decode_hex
```
Comment (LefterisJP) [inline raiden/messages.py]: style: needs trailing comma

File: raiden/messages.py [hunk 2]
```diff
@@ -200,23 +203,43 @@ class SignedMessage(Message):
...
+        data_that_was_signed = self.data_to_sign()
+        message_signature = self.signature
+
+        address = signing.recover_address(data_that_was_signed, message_signature)
+        if address is None:
+            return None
+        return address
+
+    @sender.setter
+    def sender(self, value):
```
[the remark this answers is not in the document]
Comment (LefterisJP) [reply raiden/messages.py]: yep remove this -- agree with ulo

File: raiden/network/matrixtransport.py [hunk 3]
```diff
@@ -673,27 +660,48 @@ def _sign(self, data: bytes) -> bytes:
...
+        encoded_address: str = match.group(1)
+        address: Address = to_canonical_address(encoded_address)
+        try:
+            recovered = self._recover(
+                user.user_id.encode(),
+                decode_hex(user.get_display_name()),
+            )
+            if not address or not recovered or recovered != address:
+                return
+        except Exception:  # usually, codecs raise a generic Error if no hexstr
```
[the remark this answers is not in the document]
Comment (LefterisJP) [reply raiden/network/matrixtransport.py]: yeah + 1 -- don't use the generic exception class

**Comment by ulope** [comment]
LGTM. 

Please merge after #1440 so we can take advantage of the tests.

**Pull request merged**